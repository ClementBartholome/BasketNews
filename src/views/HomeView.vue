<template>
  <SidebarFilters
    :sources="sources"
    :selectedSource="selectedSource"
    :fetchAllArticles="fetchAllArticles"
    :filterArticles="filterArticles"
  />
  <HeaderItem />
  <main>
    <BoxResults />
    <div class="home-layout">
      <span v-if="loading" class="material-symbols-outlined loader">sports_basketball</span>
      <section v-else class="articles">
        <div class="latest">
          <h2 v-if="selectedSource">Dernières actus {{ '- ' + selectedSource }}</h2>
          <h2 v-else>Dernières actus</h2>
          <article class="article latest" v-for="article in latestArticle" :key="article.link">
            <ArticleCard :article="article" />
          </article>
        </div>
        <article class="article regular" v-for="article in regularArticles" :key="article.link">
          <ArticleCard :article="article" />
        </article>
      </section>
      <router-link to="/news/page/2">
        <LoadMoreBtn>
          Afficher la suite des actus
          <span class="material-symbols-outlined"> arrow_right_alt </span></LoadMoreBtn
        ></router-link
      >
      <section class="continu">
        <h2>Chrono</h2>
        <NewsFeed :allArticles="allArticles" />
      </section>
    </div>
  </main>
</template>

<script setup lang="ts">
import { ref, onMounted } from 'vue'
import ArticleCard from '@/components/ArticleCard.vue'
import HeaderItem from '@/components/HeaderItem.vue'
import NewsFeed from '@/components/NewsFeed.vue'
import BoxResults from '@/components/BoxResults.vue'
import SidebarFilters from '@/components/SidebarFilters.vue'
import LoadMoreBtn from '@/components/LoadMoreBtn.vue'
import { sources } from '@/assets/sources'
import type { Article } from '@/types/types'
import { fetchAllArticles } from '@/api/api'

const latestArticle = ref<Article[]>([])
const regularArticles = ref<Article[]>([])
const allArticles = ref<Article[]>([])
const loading = ref(true)
const selectedSource = ref<string>('')

onMounted(async () => {
  fetchAndSortArticles()
  const intervalDuration = 60 * 1000 // 1 minute
  setInterval(fetchAndSortArticles, intervalDuration)
})

const fetchAndSortArticles = async () => {
  loading.value = true
  const fetchedArticles = await fetchAllArticles()

  fetchedArticles.sort((a, b) => b.pubDate.getTime() - a.pubDate.getTime())

  const latest = fetchedArticles.slice(0, 1)
  const regular = fetchedArticles.slice(1, 8)

  loading.value = false
  allArticles.value = fetchedArticles

  regularArticles.value = regular
  latestArticle.value = latest
  selectedSource.value = ''
}

const filterArticles = (sourceName: string) => {
  selectedSource.value = sourceName

  latestArticle.value = allArticles.value
    .filter((article) => (sourceName === '' ? true : article.source === sourceName))
    .slice(0, 1)
  regularArticles.value = allArticles.value
    .filter((article) => (sourceName === '' ? true : article.source === sourceName))
    .slice(1, 8)

  if (selectedSource.value === '') {
    fetchAndSortArticles()
  }
}
</script>

<style lang="scss">
.dark-mode {
  a {
    color: var(--text-color);
  }
}

main {
  max-width: 1280px;
  width: 100%;
  margin: 0 auto;
  grid-row: 2;
  grid-column: 2;

  @media (max-width: 768px) {
    width: 100%;
  }
}

h2 {
  margin: 10px 0;
}

p {
  margin: 5px 0;
}

a {
  text-decoration: none;
  color: var(--text-color-light);
}

button {
  background: none;
  border: none;
  cursor: pointer;
}

.articles {
  gap: 1.5rem;
  display: flex;
  flex-wrap: wrap;

  @media (max-width: 425px) {
    margin-right: 0;
  }

  a:hover h2 {
    color: #007bff;
  }

  .regular {
    margin-top: 52px;

    .article-title {
      font-size: 1.2rem;
    }

    p {
      font-size: 0.9rem;
    }

    @media (max-width: 1200px) {
      .article-title {
        font-size: 1.4rem;
      }

      p {
        font-size: 1.2rem;
      }
    }
  }
}

.article.latest {
  max-width: 500px;

  @media (max-width: 1200px) {
    max-width: 100%;
    width: 100%;
  }

  .article-title {
    font-size: 1.4rem;
  }

  p {
    font-size: 1.2rem;
  }
  .img-zoom {
    height: 300px;
  }
}

.continu {
  background-color: var(--background-secondary);
  padding: 15px;
  border-radius: 10px;
  margin-top: 10px;
  margin-left: 20px;
  height: fit-content;
  grid-row: 1;
  grid-column: 2;

  @media (max-width: 900px) {
    grid-row: 3;
    grid-column: 1;
  }
}

.home-layout {
  display: grid;
  grid-template-columns: 1fr 320px;
  margin: 10px;
}

@media (max-width: 1200px) {
  .article,
  .latest {
    max-width: 100%;

    .img-zoom {
      height: 300px;
    }
  }
}

@media (max-width: 900px) {
  .home-layout {
    grid-template-columns: 1fr;
    margin: 20px;
  }
  .continu {
    margin: 30px auto;
  }
}

@media (max-width: 850px) {
  .article,
  .latest,
  .regular {
    width: 100%;
    max-width: none;
  }
  .article .img-zoom {
    height: 320px;
  }
}

@media (max-width: 425px) {
  main {
    padding: 10px;
  }
  .home-layout {
    margin: 0;
  }
}

@media (max-width: 375px) {
  .articles {
    margin-right: 0;
  }
}

.loader {
  position: relative;
  animation: bouncy-loader 0.6s infinite alternate;
  font-size: 5rem;
  margin: 0 auto;
}
@keyframes bouncy-loader {
  0% {
    top: 0;
  }
  33% {
    top: 20px;
  }
  66% {
    top: 0;
  }
  100% {
    top: 20px;
  }
}
</style>
