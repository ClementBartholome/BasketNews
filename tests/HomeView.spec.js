import { mount } from '@vue/test-utils'
import { expect, describe, it } from 'vitest'
import HomeView from '../src/views/HomeView.vue'

describe('HomeView', () => {
  it('sorts articles correctly and updates values', async () => {
    const wrapper = mount(HomeView)

    // Call the function to fetch and sort articles
    await wrapper.vm.fetchAndSortArticles()

    // Verify that loading is false after sorting
    expect(wrapper.vm.loading).toBe(false)

    // Verify that the total number of articles is 80
    expect(wrapper.vm.allArticles).toHaveLength(80)
  })

  it('filters articles correctly', async () => {
    const wrapper = mount(HomeView)

    // Initialize values before calling filterArticles
    wrapper.vm.selectedSource = ''
    wrapper.vm.latestArticle = []
    wrapper.vm.regularArticles = []

    // Call the function to fetch and sort articles
    await wrapper.vm.fetchAndSortArticles()

    // Call the function you want to test
    wrapper.vm.filterArticles('BasketUSA')

    // Verify that selectedSource has been updated
    expect(wrapper.vm.selectedSource).toBe('BasketUSA')

    // Verify that latestArticle and regularArticles have been updated correctly
    expect(wrapper.vm.latestArticle).toHaveLength(1)
    expect(wrapper.vm.regularArticles).toHaveLength(7)

    // Verify that fetchAndSortArticles is called when selectedSource is empty
    wrapper.vm.filterArticles('')
  })
})
