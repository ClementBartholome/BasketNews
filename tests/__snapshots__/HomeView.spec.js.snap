// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeView > renders correctly 1`] = `
"<aside data-v-f44fcf8d=\\"\\" class=\\"\\">
  <div data-v-f44fcf8d=\\"\\" class=\\"logo\\"></div>
  <div data-v-f44fcf8d=\\"\\" class=\\"menu-toggle-wrap\\"><button data-v-f44fcf8d=\\"\\" class=\\"menu-toggle\\"><span data-v-f44fcf8d=\\"\\" class=\\"material-symbols-outlined\\">keyboard_double_arrow_right</span></button></div>
  <h3 data-v-f44fcf8d=\\"\\">Menu</h3>
  <div data-v-f44fcf8d=\\"\\" class=\\"filter-buttons menu\\"><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button active\\" title=\\"Voir tous les articles\\"><span data-v-f44fcf8d=\\"\\" class=\\"material-symbols-outlined\\">sports_basketball</span><span data-v-f44fcf8d=\\"\\" class=\\"text\\"> Tous</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"BasketUSA\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/basketusa.webp\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">BasketUSA</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"TrashTalk\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/trashtalk.png\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">TrashTalk</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"BasketSession\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/basketsession.webp\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">BasketSession</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"BasketEurope\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/basketeurope.webp\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">BasketEurope</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"L'Equipe\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/lequipe.webp\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">L'Equipe</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"ESPN NBA\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/espn.png\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">ESPN NBA</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"CBS Sports NBA\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/cbs.webp\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">CBS Sports NBA</span></button><button data-v-f44fcf8d=\\"\\" class=\\"filter-button button\\" title=\\"New York Times\\"><img data-v-f44fcf8d=\\"\\" src=\\"https://raw.githubusercontent.com/ClementBartholome/BasketNews/main/public/images/nytimes.png\\" class=\\"material-symbols-outlined logo\\" alt=\\"Logo\\"><span data-v-f44fcf8d=\\"\\" class=\\"text\\">New York Times</span></button></div>
</aside>
<header data-v-0a0c25dd=\\"\\" class=\\"header\\">
  <router-link data-v-0a0c25dd=\\"\\" to=\\"/\\">
    <div data-v-0a0c25dd=\\"\\" class=\\"header-content\\"><span data-v-0a0c25dd=\\"\\" class=\\"material-symbols-outlined\\"> sports_basketball </span>
      <h1 data-v-0a0c25dd=\\"\\">Daily Basket</h1>
    </div>
  </router-link><button data-v-31be2b8d=\\"\\" data-v-0a0c25dd=\\"\\" class=\\"dark-mode-btn\\"><span data-v-31be2b8d=\\"\\" class=\\"material-symbols-outlined\\">light_mode</span></button>
</header>
<main>
  <section data-v-841a08ad=\\"\\" class=\\"box-results\\"><span data-v-841a08ad=\\"\\" class=\\"yesterday\\">NBA<br>Hier</span><span data-v-841a08ad=\\"\\" class=\\"no-games\\">Pas de matchs NBA jou√©s hier</span><button data-v-841a08ad=\\"\\" class=\\"change-day\\"><span data-v-841a08ad=\\"\\" class=\\"material-symbols-outlined\\">arrow_forward</span></button></section>
  <div class=\\"home-layout\\"><span class=\\"material-symbols-outlined loader\\">sports_basketball</span>
    <router-link to=\\"/news/page/2\\"><button data-v-b2ec4738=\\"\\"> Afficher la suite des actus <span class=\\"material-symbols-outlined\\"> arrow_right_alt </span></button></router-link>
    <section class=\\"continu\\">
      <h2>Chrono</h2>
      <section data-v-a404d4cc=\\"\\"><button data-v-b2ec4738=\\"\\" data-v-a404d4cc=\\"\\">Afficher les actus NBA suivantes <span data-v-a404d4cc=\\"\\" class=\\"material-symbols-outlined\\">arrow_drop_down</span></button></section>
    </section>
  </div>
</main>"
`;
